### PR TITLE
chore: refactor training loop

### PR DESCRIPTION
**Pull request by caic99**
## Summary by CodeRabbit

- **New Features**
	- Enhanced training loop to support multi-task training, allowing for more flexible model selection.

- **Improvements**
	- Streamlined `step` function to accept only the step ID, simplifying its usage.
	- Adjusted logging and model saving mechanisms for consistency with the new training flow.
	- Improved random seed management for enhanced reproducibility in data processing.
	- Enhanced error handling in data retrieval to ensure seamless operation during data loading.
	- Added type hints for better clarity in data loader attributes.

### CODE REVIEW

File: deepmd/pt/train/training.py [hunk 1]
```diff
@@ -653,6 +653,12 @@ def run(self) -> None:
             prof.start()
 
         def step(_step_id, task_key="Default") -> None:
+            if self.multi_task:
+                model_index = dp_random.choice(
+                    np.arange(self.num_model),
```
Comment (caic99) [inline deepmd/pt/train/training.py:658]:
```suggestion
                    np.arange(self.num_model, dtype=np.int_),
```

File: deepmd/pt/train/training.py [hunk 2]
```diff
@@ -653,6 +653,12 @@ def run(self) -> None:
             prof.start()
 
         def step(_step_id, task_key="Default") -> None:
+            if self.multi_task:
+                model_index = dp_random.choice(
+                    np.arange(self.num_model, dtype=np.int_),
+                    p=self.model_prob,
+                )
```
Comment (njzjz) [inline deepmd/pt/train/training.py:660]: I see the hack for multiprocessing is removed. Have we ensured random seeds are different in different processes?

Comment (njzjz) [reply deepmd/pt/train/training.py:660]: See also https://numpy.org/doc/stable/reference/random/parallel.html

Comment (caic99) [reply deepmd/pt/train/training.py:660]: > Have we ensured random seeds are different in different processes?

Yes, by not setting the seed explicitly.

https://github.com/deepmodeling/deepmd-kit/blob/e695a91ca6f7a1c9c830ab1c58b7b7a05db3da23/deepmd/utils/random.py#L9
This initializes the generator with a random state.
For torch, the `seed()` method is never called.
For tensorflow it goes here
https://github.com/deepmodeling/deepmd-kit/blob/e695a91ca6f7a1c9c830ab1c58b7b7a05db3da23/deepmd/tf/entrypoints/train.py#L220-L225

Comment (njzjz) [reply deepmd/pt/train/training.py:660]: If the seed is not set explicitly, different processes may have the same seed.

Comment (njzjz) [reply deepmd/pt/train/training.py:660]: TF seed is also wrong.

Comment (caic99) [reply deepmd/pt/train/training.py:660]: > If the seed is not set explicitly, different processes may have the same seed.

`[np.random.RandomState](https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState)` initializes seed from random generators if the seed is not provided. So I think it would be less likely for different processes having the same seed.

Comment (caic99) [reply deepmd/pt/train/training.py:660]: @njzjz Do you think we should mandatorily set a seed?

Comment (njzjz) [reply deepmd/pt/train/training.py:660]: > @njzjz Do you think we should mandatorily set a seed?

Indeed, it's surprising that the random seed set by the user is not used, meaning the training may not be reproducible.

File: deepmd/pt/utils/dataloader.py [hunk 3]
```diff
@@ -50,6 +51,7 @@ def setup_seed(seed) -> None:
     torch.manual_seed(seed)
     torch.cuda.manual_seed_all(seed)
     torch.backends.cudnn.deterministic = True
+    dp_random.seed(seed)
```
Comment (njzjz) [inline deepmd/pt/utils/dataloader.py:54]: should also seed with rank, i.e. [rank, seed]; see the example at https://numpy.org/doc/stable/reference/random/parallel.html#sequence-of-integer-seeds.

Comment (caic99) [reply deepmd/pt/utils/dataloader.py:54]: Maybe in another PR fixing #4440?

Comment (caic99) [reply deepmd/pt/utils/dataloader.py:54]: Currently the seed has been set to _seed from config file_ + rank, which is OK for the basic need.

Comment (njzjz) [reply deepmd/pt/utils/dataloader.py:54]: > Currently the seed has been set to seed from config file + rank

I don't get this point. It's done by TF but not PT.

Comment (caic99) [reply deepmd/pt/utils/dataloader.py:54]: > Currently the seed has been set to _seed from config file_ + rank, which is OK for the basic need.

@njzjz here
https://github.com/deepmodeling/deepmd-kit/blob/e695a91ca6f7a1c9c830ab1c58b7b7a05db3da23/deepmd/pt/entrypoints/main.py#L141

**Review by njzjz — APPROVED** [review]
